wgsl/index.bs

```diff
@@ -600,28 +642,334 @@ storage_class
+The memory layout of a type is significant only when evaluating an expression with:
+* a variable in the `uniform` or `storage` storage class, or
+* a pointer into the `uniform` or `storage` storage class.
+
+An 8-bit byte is the most basic unit of [=host-shareable=] memory.
+The terms defined in this section express counts of 8-bit bytes.
+
+We will use the following notation:
+
+* |Stride|(|A|) is the value of the [=stride=] attribute of array type |A|.
+* |Offset|(|S|,|i|) is the value of the [=offset=] attribute of the |i|'th member of structure type |S|.
+
+The remainder of this section is structured as follows:
+* Section [[#memory-layout-intent]] describes the intent of the layout rules.  It is not normative.
+* Section [[#internal-value-layout]] says how the internals of a value are placed in byte memory
+    locations starting from a given byte offset in a buffer.  For composite types the layout
+    depends on the properties of the type itself, the storage class, and the user-specified
+    [=stride=] and [=offset=] attributes.
+* Section [[#layout-constraints]] describes the contraints on array strides and structure member
+    offsets, ultimately yielding definitions for
+    [=uniform buffer layout=] and [=storage buffer layout=].
+
+#### Memory Layout Intent #### {#memory-layout-intent}
+
+This section is informative, not normative.
+
+The layout rules describe two sets of constraints, one for uniform buffers and one for storage buffers.
+They are similar in many respects, but the uniform buffer layout is more restrictive.
+
+In particular:
+* Scalar data values are aligned to their own size.
+* There is no padding between components of a vector.
+* Two-element and four-element vectors are aligned to their size.
+* Three-element vectors are aligned as if they were four-element vectors.
+* An array's element alignment is a multiple of the element type's alignment.
+    * In a uniform buffer, the array element alignment is also a multiple of 16.
+* An array's alignment is the same as its element alignment.
+* A matrix with |N| columns is aligned as if it were an array of |N| column vectors.
+* A structure inherits the worst-case alignment of any of its members.
+    * In a uniform buffer, structure member alignment is also a multiple of 16.
+* The members of a structure are laid out in order: earlier members are appear earlier in
+    the buffer
+
+Additionally we define a value's allocation extent, or memory footprint, which determines
+how many memory locations must be reserved to store that value in host-shareable memory.
+Allocation extent is a determining factor of the minimum size of a buffer that can be bound
+to a uniform buffer variable
+or to a storage buffer variable.
+See [[#resource-layout-compatibility]].
+
+Compared to OpenGL:
+* OpenGL `std140` layout satisfies [SHORTNAME] [=uniform buffer layout=].
```

Comment:
is it "satisfies", or is it "exactly matches"?

Reply:
Satisfies.  WGSL uniform buffer layout can give strides and offsets that are multiples of the ones specified by std140.

Example:

      uint arr[3];   // will have stride of 16 in std140. 

But 

      [[stride(32)]] array<u32,3>; // This is perfectly fine to use inside a uniform buffer in WGSL

Reply:
Can we instead say something like
> For any type |S|, the tightest offset and stride assignments would exactly match both [SHORTNAME] [=uniform buffer layout=] and `std140`

Reply:
This is unresolved, right?

Reply:
How about:

* For any type |S|, OpenGL std140 layout is the same as using the tightest offset and stride assignments in [SHORTNAME] [=uniform buffer layout=].

Reply:
Or remove the quantification by |S|:

*  For any type, OpenGL `std140` layout is the same as using the tightest offset and stride assignments in WGSL [=uniform buffer layout=].